[Merged by Bors] - feat(data/list/basic): Add lemmas about inits and tails

eric-wieser opened the pull request:
---
<!-- put comments you want to keep out of the PR commit here -->

eric-wieser committed:
feat(data/list/basic): Add lemmas about inits and tails

eric-wieser committed:
feat(data/list/zip): Add lemmas about inits and tails

eric-wieser committed:
feat(data/list/basic,logic/function/basic): Add list.reverse_involutive.comp_self

eric-wieser committed:
feat(data/list/basic): Add lemmas relating inits and tails

eric-wieser commented on src/data/list/basic.lean:
```diff
@@ -2975,6 +2978,40 @@ instance decidable_suffix [decidable_eq α] : ∀ (l₁ l₂ : list α), decidab
+by { rw ← reverse_map_reverse_inits_reverse l, simp [reverse_involutive.comp_self], }
+
+lemma tails_reverse (l : list α) : tails (reverse l) = map reverse (reverse l.inits) :=
+by { rw ← reverse_map_reverse_tails_reverse l, simp [reverse_involutive.comp_self], }
```
I'm not sure whether all four of these are sensible, or if I should just keep one pair. `reverse` and `map reverse` are involutive, so it's a question of whether they should appear on the left or the right of the lemma. I suppose then I could also state this
```
tails ∘ reverse = reverse ∘ map reverse ∘ inits
```
I suppose it comes down to what is easiest to rewrite with, and I don't have a strong use-case yet - primarily it just seems neat to show how they relate to each other.

Vierkantor replied:
It makes sense to keep both: sometimes you want to talk about `inits l` but you only have `tails`, and sometimes you want to talk about `inits (reverse l)`.

eric-wieser replied:
The problem with that logic is that there is an explosion of possible orderings, all of which are trivial to obtain from the others (`x` marks the ones in this PR):
* [x] `tails = reverse ∘ map reverse ∘ inits ∘ reverse`
* [ ] `tails = map reverse ∘ reverse ∘ inits ∘ reverse`
* [ ] `tails ∘ reverse = reverse ∘ map reverse ∘ inits`
* [x] `tails ∘ reverse = map reverse ∘ reverse ∘ inits`
* [ ] `map reverse ∘ tails = reverse ∘ inits ∘ reverse`
* [ ] `reverse ∘ map reverse ∘ tails = inits ∘ reverse`
* [x] `map reverse ∘ reverse ∘ tails = inits ∘ reverse`
* [x] `reverse ∘ map reverse ∘ tails ∘ reverse = inits`
* [ ] `map reverse ∘ reverse ∘ tails ∘ reverse = inits`

All of these follow trivially from `involutive reverse` and `involutive (map reverse)`, and a proof of one of them.

Vierkantor replied:
Good point. But `map reverse ∘ reverse` gets `simp`ed to `reverse ∘ map reverse`, so taking those out we get a more manageable list:

* [x] `tails = reverse ∘ map reverse ∘ inits ∘ reverse`
* [ ] `tails ∘ reverse = reverse ∘ map reverse ∘ inits` (so we should probably change `tails ∘ reverse = map reverse ∘ reverse ∘ inits` into this one)
* [ ]  `reverse ∘ tails ∘ reverse = map reverse ∘ inits` (I believe you missed this one)
* [ ] `map reverse ∘ tails = reverse ∘ inits ∘ reverse`
* [ ] `reverse ∘ map reverse ∘ tails = inits ∘ reverse` (idem for `map reverse ∘ reverse ∘ tails = inits ∘ reverse`)
* [x] `reverse ∘ map reverse ∘ tails ∘ reverse = inits`

And honestly including `reverse ∘ tails ∘ reverse = map reverse ∘ inits` (or the symmetric version) and its `inits`-dual doesn't sound too bad: with 6 lemmas we cover every combination of `tails`, `inits`, `reverse` and `map_reverse`.

eric-wieser replied:
Should I stick to the applied versions of the lemmas, or use the `∘` forms we're discussing here?

Vierkantor replied:
Applied should be fine. (We should really make a tactic that derives one set of lemmas from the other!)

eric-wieser replied:
Will add these once the merge build finishes and I have a cache to work from

eric-wieser replied:
Added the six lemmas you suggested, and tweaked the names to match other lemmas in the file

eric-wieser commented on src/data/list/zip.lean:
```diff
@@ -138,4 +138,18 @@ by rw [← zip_unzip.{u u} (revzip l).reverse, unzip_eq_map]; simp; simp [revzip
+  obtain (rfl | ⟨i', h, rfl⟩) := hi,
+  simp,
+  simp [l_ih i' h],
+end
```
Will update this to be `iff`, and fix some unwanted explicit parameters

jcommelin commented on src/data/list/basic.lean:
```diff
@@ -2975,6 +2978,40 @@ instance decidable_suffix [decidable_eq α] : ∀ (l₁ l₂ : list α), decidab
+lemma tails_cons (a : α) (l : list α) : tails (a :: l) = (a :: l) :: l.tails :=
+by simp
+
+@[simp]
```
Is the RHS really simpler than the LHS? Similar for the lemma below this one.

eric-wieser replied:
I don't feel strongly. I figured if `*_cons` is `simp`, the inductive nature of `++` means that `*_append` should be too. Is there an existing counterexample to that claim?

Vierkantor replied:
They seem useful enough to be `simp` lemmas. I think the following looks nicer and matches `tails_append` better though.
```lean
lemma inits_append : ∀ (s t : list α), inits (s ++ t) = s.inits.init ++ t.inits.map (λ l, s ++ l)
```

eric-wieser replied:
I was not aware of `.init`, but I agree the symmetry is nice

eric-wieser replied:
However, `.tail` tends to be easier to work with inductively I think

Vierkantor replied:
That's true. Let's stick to `tail` then.

gebner commented on src/data/list/basic.lean:
```diff
@@ -552,8 +552,11 @@ by rw [concat_eq_append, reverse_append, reverse_singleton, singleton_append]
 @[simp] theorem reverse_reverse (l : list α) : reverse (reverse l) = l :=
 by induction l; [refl, simp only [*, reverse_cons, reverse_append]]; refl
 
+theorem reverse_involutive : involutive (@reverse α) :=
```
```suggestion
@[simp] theorem reverse_involutive : involutive (@reverse α) :=
```

eric-wieser replied:
Should _every_ involutive statement in mathlib be tagged `simp`?

gebner commented on src/data/list/basic.lean:
```diff
@@ -552,8 +552,11 @@ by rw [concat_eq_append, reverse_append, reverse_singleton, singleton_append]
+theorem reverse_involutive : involutive (@reverse α) :=
+λ l, reverse_reverse l
+
 theorem reverse_injective : injective (@reverse α) :=
```
```suggestion
@[simp] theorem reverse_injective : injective (@reverse α) :=
```

eric-wieser replied:
This change causes a CI failure with:
```
/- Checking 51416 declarations (plus 50261 automatically generated ones) in mathlib -/

/- The `simp_var_head` linter reports: -/
/- LEFT-HAND SIDE HAS VARIABLE AS HEAD SYMBOL.
Some simp lemmas have a variable as head symbol of the left-hand side: -/
-- data/list/basic.lean
#print list.reverse_injective /- Left-hand side has variable as head symbol: a₁ -/
```

gebner replied:
I believe this is a bug in the simp linter.  Just leave it out of the PR.  I'll try to add it in another PR with a fix to the linter.

gebner replied:
#4237

eric-wieser committed:
chore(data/list/basic): Tag reverse theorems with simp

Co-authored-by: Gabriel Ebner <gebner@gebner.org>

eric-wieser committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into list-lemmas

eric-wieser commented on src/data/list/basic.lean:
```diff
@@ -552,8 +552,11 @@ by rw [concat_eq_append, reverse_append, reverse_singleton, singleton_append]
+@[simp] theorem reverse_involutive : involutive (@reverse α) :=
+λ l, reverse_reverse l
+
+@[simp] theorem reverse_injective : injective (@reverse α) :=
```
```suggestion
theorem reverse_injective : injective (@reverse α) :=
```

eric-wieser committed:
revert(data/list/basic): Do not mark reverse_injective with simp

This failed the linter with the message
```
/- The `simp_var_head` linter reports: -/
/- LEFT-HAND SIDE HAS VARIABLE AS HEAD SYMBOL.
Some simp lemmas have a variable as head symbol of the left-hand side: -/
-- data/list/basic.lean
#print list.reverse_injective /- Left-hand side has variable as head symbol: a₁ -/
```

Vierkantor commented on src/data/list/zip.lean:
```diff
@@ -167,4 +167,23 @@ begin
+  simp_rw [tails, inits, zip_cons_cons],
+  { simp },
+  { split;
+    rw [mem_cons_iff, zip_map_left, mem_map, prod.exists],
```
We treat `;` as if it's an operator, not a separator:
```suggestion
  { split; rw [mem_cons_iff, zip_map_left, mem_map, prod.exists],
```

Vierkantor commented on src/data/list/zip.lean:
```diff
@@ -167,4 +167,23 @@ begin
+  (init, tail) ∈ zip l.inits l.tails ↔ init ++ tail = l :=
+begin
+  induction l generalizing init tail;
+  simp_rw [tails, inits, zip_cons_cons],
```
```suggestion
  induction l generalizing init tail;
    simp_rw [tails, inits, zip_cons_cons],
```

eric-wieser committed:
style(data/list/zip): whitespace fixes

Co-authored-by: Vierkantor <Vierkantor@users.noreply.github.com>

eric-wieser committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into list-lemmas

eric-wieser committed:
chore(data/list/basic): Rename tails/init lemmas, and add the missing two from the set

Vierkantor reviewed (approved):
Looks good to me. I'll give the previous reviewers the chance to comment before we merge.

jcommelin commented on src/data/list/zip.lean:
```diff
@@ -167,4 +167,22 @@ begin
   { rintro ⟨h₀, h₁⟩, exact ⟨_,_,h₀,h₁,rfl⟩ }
 end
 
+lemma zip_inits_tails {l : list α} {init tail : list α} :
```
```suggestion
lemma mem_zip_inits_tails {l : list α} {init tail : list α} :
```

eric-wieser replied:
Good idea, fixed

jcommelin reviewed (commented):
lgtm

eric-wieser committed:
Update src/data/list/zip.lean

Co-authored-by: Johan Commelin <johan@commelin.net>
